### ISSUE TITLE
Network.fromJSON / .toJSON is incomplete

### ISSUE THREAD

**Issue by christianechevarria** [liquidcarrot/carrot#152]
### Description
In theory, `Network.fromJSON(Network.toJSON())` should produce a deep copy of a network but as noted in #149 this is not the case. It seems that the import / export does not currently involve all of details necessary to achieve this.

### Screenshots
![Network Import Export Bug](https://user-images.githubusercontent.com/23618650/62843588-e67c7100-bc88-11e9-8bcf-386127856595.png)

Dropout is a top level parameter that is not included in the import / export system and should

### Files
Network.js

### Extra considerations
We should specifically consider making this import / export system compatible with other more popular libraries like Tensorflow for maximum interoperability

**Comment by christianechevarria** [comment]
Update:

Although we've solved the #149 using Lodash' `.deepClone` method, this is a MASSIVE bottleneck in the library and is seriously affecting performance -- increasing our test times by as much as 3x

For the sake of performance we need to make sure we can create deep copies of Networks that are indistinguishable from the original using our custom `.toJSON` & `.fromJSON` system

**Comment by christianechevarria** [comment]
Update: 

Thinking that the best approach to this is to create a faithful neuron level .fromJSON / .toJSON and just call that repeatedly when doing fromJSON / toJSON at higher levels of abstraction like the Network

Focusing efforts there for now

**Comment by christianechevarria** [comment]
@tracycollins @dan-ryan @luiscarbonell would any of you know of a good tool to profile JS code to see what the fastest / slowest parts (of the code) are of a function's execution?

It would be great to validate just how much of a slow-down using Lodash's `deepClone` is and see the pay-off of solving the fromJSON / toJSON issue

**Comment by tracycollins** [comment]
> @tracycollins @dan-ryan @luiscarbonell would any of you know of a good tool to profile JS code to see what the fastest / slowest parts (of the code) are of a function's execution?
> 
> It would be great to validate just how much of a slow-down using Lodash's `deepClone` is and see the pay-off of solving the fromJSON / toJSON issue

The only profiling tool that I've used is ndb:  https://www.npmjs.com/package/ndb

Don't know how it compares to other tools (cuz I've not used anything else), but it's helped me when I needed it.

**Comment by tracycollins** [comment]
Any progress on this? If not, is there a workaround?

I need to be able to store a network in a database, and so have to use the toJSON() method. When I attempt to use fromJSON(), it (usually?) fails with an error: "No input or output size given"

**Comment by christianechevarria** [comment]
@tracycollins I was stalled trying to deep copy arbitrary functions on the node level, but I can probably get this working pretty quick with the built in functions -- are you trying to do the toJSON at the Network level?

**Comment by christianechevarria** [comment]
I'll focus on this and get a hot patch in

**Comment by tracycollins** [comment]
@christianechevarria Yes, using Network.toJSON() and Network.fromJSON().  It'd be great if you can get this done so quickly. Thanks!

**Comment by raimannma** [comment]
I almost had the same issue it's a quick change in one toJSON() method, where the JSON keys are different between toJSON and fromJSON

**Comment by christianechevarria** [comment]
@raimannma just pushed up a branch called `fix-network-json-system` 

Also, out of curiosity was the fix related to this code?

https://github.com/liquidcarrot/carrot/blob/02c90caadcdd623ebc94e6f703362bb7613ae4b8/src/architecture/network.js#L1604-L1607

Changed it to use the old (better?) `.input` and `.output` naming

```
Network.fromJSON = function(json) {
  // TODO: Match new network input/output nodes with json.input nodes and json.output nodes

  const network = new Network(json.input, json.output); // Use JSON input / output for backwards compatibility
```

**Comment by christianechevarria** [comment]
@tracycollins @raimannma and I were having issues reproducing, would you be able to tell us what version you're running and what the code looks like to have a better idea what's going on?

**Comment by tracycollins** [comment]
> @tracycollins @raimannma and I were having issues reproducing, would you be able to tell us what version you're running and what the code looks like to have a better idea what's going on?

Looks like the problem I was having was due to my using both carrot and neataptic libraries, and trying to erroneously convert networks between the two. This was a bug in my code.

Although this does not seem to be possible now, might this be possible (and useful) in the future? And not just between carrot and neataptic, but between other libraries as well? Is there anything like a generic/standard JSON format for networks?

**Comment by christianechevarria** [comment]
> > @tracycollins @raimannma and I were having issues reproducing, would you be able to tell us what version you're running and what the code looks like to have a better idea what's going on?
> 
> Looks like the problem I was having was due to my using both carrot and neataptic libraries, and trying to erroneously convert networks between the two. This was a bug in my code.
> 
> Although this does not seem to be possible now, might this be possible (and useful) in the future? And not just between carrot and neataptic, but between other libraries as well? Is there anything like a generic/standard JSON format for networks?

@teh-mICON and I were just discussing this in the context of building a generic neural-network visualization library that is inter-operable with all JS neural-network libraries, but as far as I'm aware there's no standard to date for JS specifically -- which is a big bummer lol. 

There is [ONNX](https://github.com/onnx/onnx) though, which is popular in python and was already requested by @loaiabdalslam on #159, but to be honest I'm not sure on how to begin working on this one, but I'm sure there's someone in the open source community who would -- just not sure who they are yet.

Another idea that could work is potentially replicating a system like ONNX in JS and maybe making it inter-operable with python modes in the future as our community's knowledge increases or new tech is developed (this is something that should be added to the discussion in #178 since WASM could be a huge resource for getting it done across libraries and languages).

As far as getting a standard adopted, there's an ML group in JS called [BRIIM](https://bri.im/) which includes [Brain.js](https://github.com/BrainJS/brain.js/pull/190) and [Neataptic](https://github.com/wagenaartje/neataptic/pull/133) which might be a good place to start proposing an idea like this. It seems like both were spearheaded by @rwieruch so maybe we can start by approaching him

**Comment by tracycollins** [comment]
Having a generic/standard would be a big help when evaluating different libraries. I'll take a look at the links.  Thanks

**Comment by GavinRay97** [comment]
@christianechevarria ONNX.js exists, it's by Microsoft:

https://github.com/microsoft/onnxjs

It has the same internal architecture for performance that TF.js does, if you check the `onnxjs/lib/backends` directory, they use the same dynamic backend pattern where it picks either WebGL + shader hacks to use GPU or WASM to use CPU. Pretty neat IMO

**Comment by rwieruch** [comment]
> As far as getting a standard adopted, there's an ML group in JS called [BRIIM](https://bri.im/) which includes [Brain.js](https://github.com/BrainJS/brain.js/pull/190) and [Neataptic](https://github.com/wagenaartje/neataptic/pull/133) which might be a good place to start proposing an idea like this. It seems like both were spearheaded by @rwieruch so maybe we can start by approaching him

Thanks for the shoutout @christianechevarria 👍 

The website wasn't updated after its initial launch, but the Slack Community is quite active https://slack.bri.im In the beginning I made sure to invite all the library contributors of the JS + ML ecosystem. Every library has its own channel 🙂

**Comment by christianechevarria** [comment]
> @tracycollins I was stalled trying to deep copy arbitrary functions on the node level, but I can probably get this working pretty quick with the built in functions -- are you trying to do the toJSON at the Network level?

Update: found a [handy little function](https://github.com/vkiryukhin/jsonfn) for deep copying objects that preserve simple pure functions (like the ones we'd need for activation functions), will get integrated into the fromJSON / toJSON system

**Comment by christianechevarria** [comment]
> Fixed in typescript version

We still have the `Node` reference in x_traces, but once we rework it then it would be good

**Comment by raimannma** [comment]
@christianechevarria no, i removed that. So json system should be finished
